File: test_samples/all_languages/UserService.scala
// Complex Scala file with advanced features
package com.example.app

import scala.concurrent.{Future, ExecutionContext}
import scala.util.{Try, Success, Failure}

// Case classes
case class User(id: Long, username: String, email: String)
case class Post(id: Long, userId: Long, title: String, content: String)

// Sealed trait (similar to sealed class in Kotlin)
sealed trait Result[+A]
case class Ok[A](value: A) extends Result[A]
case class Err(message: String) extends Result[Nothing]
case object Loading extends Result[Nothing]

// Trait with type parameters
trait Repository[T] {
  def findById(id: Long): Future[Option[T]]
  def findAll(): Future[Seq[T]]
  def save(entity: T): Future[T]
}

// Implicit parameters and type classes
trait JsonSerializer[T] {
  def toJson(value: T): String
}

object JsonSerializer {
  implicit val userSerializer: JsonSerializer[User] = new JsonSerializer[User] {
    def toJson(user: User): String = 
      s"""{"id":${user.id},"username":"${user.username}","email":"${user.email}"}"""
  }
}

// Generic class with implicit evidence
class UserRepository(implicit ec: ExecutionContext) extends Repository[User] {
  private var users = Map.empty[Long, User]
  
  override def findById(id: Long): Future[Option[User]] = Future {
    users.get(id)
  }
  
  override def findAll(): Future[Seq[User]] = Future {
    users.values.toSeq
  }
  
  override def save(user: User): Future[User] = Future {
    users = users + (user.id -> user)
    user
  }
}

// Pattern matching and for-comprehensions
class UserService(repository: Repository[User])(implicit ec: ExecutionContext) {
  def getUserWithPosts(userId: Long): Future[Result[(User, Seq[Post])]] = {
    val result = for {
      userOpt <- repository.findById(userId)
      user <- Future.successful(userOpt)
      // Simulated post retrieval
      posts = Seq.empty[Post]
    } yield (user, posts)
    
    result.map {
      case Some(data) => Ok(data)
      case None => Err("User not found")
    }.recover {
      case ex: Exception => Err(ex.getMessage)
    }
  }
  
  // Higher-order functions
  def processUsers(predicate: User => Boolean): Future[Seq[User]] = {
    repository.findAll().map(_.filter(predicate))
  }
}

// Companion object with apply method
object UserService {
  def apply(repository: Repository[User])(implicit ec: ExecutionContext): UserService = {
    new UserService(repository)
  }
}

// Type aliases and implicit conversions
type UserId = Long
type Username = String

object Converters {
  implicit class UserOps(user: User) {
    def toJson(implicit serializer: JsonSerializer[User]): String = {
      serializer.toJson(user)
    }
  }
}

// Partial functions
object UserValidator {
  val isValidEmail: PartialFunction[User, Boolean] = {
    case User(_, _, email) if email.contains("@") => true
  }
  
  val isValidUsername: PartialFunction[User, Boolean] = {
    case User(_, username, _) if username.length >= 3 => true
  }
  
  val validate: PartialFunction[User, Boolean] = 
    isValidEmail orElse isValidUsername
}

File: test_samples/all_languages/UserService.lua
-- Complex Lua script with advanced features

-- Module definition
local UserService = {}
UserService.__index = UserService

-- Constructor
function UserService:new(repository)
    local instance = setmetatable({}, UserService)
    instance.repository = repository
    instance.cache = {}
    return instance
end

-- Metatable for User
local User = {}
User.__index = User

function User:new(id, username, email)
    return setmetatable({
        id = id,
        username = username,
        email = email,
        created_at = os.time()
    }, User)
end

function User:validate()
    if not self.username or #self.username < 3 then
        return false, "Username too short"
    end
    if not self.email or not self.email:match("@") then
        return false, "Invalid email"
    end
    return true
end

-- Class methods
function UserService:findById(id)
    -- Check cache first
    if self.cache[id] then
        return self.cache[id]
    end
    
    -- Fetch from repository
    local user = self.repository:findById(id)
    if user then
        self.cache[id] = user
    end
    return user
end

function UserService:findAll(options)
    options = options or {}
    local limit = options.limit or 100
    local offset = options.offset or 0
    
    return self.repository:findAll(limit, offset)
end

-- Higher-order function
function UserService:map(users, fn)
    local result = {}
    for i, user in ipairs(users) do
        table.insert(result, fn(user))
    end
    return result
end

-- Coroutine example
function UserService:processAsync(userId, callback)
    local co = coroutine.create(function()
        local user = self:findById(userId)
        if user then
            callback(nil, user)
        else
            callback("User not found")
        end
    end)
    
    return coroutine.resume(co)
end

-- Closure example
function UserService:createValidator(minLength)
    return function(username)
        return #username >= minLength
    end
end

-- Table manipulation
local function deepCopy(orig)
    local copy
    if type(orig) == 'table' then
        copy = {}
        for k, v in pairs(orig) do
            copy[deepCopy(k)] = deepCopy(v)
        end
    else
        copy = orig
    end
    return copy
end

-- Module exports
return {
    UserService = UserService,
    User = User,
    deepCopy = deepCopy
}
